pkg/eigenState/stakerShares/stakerShares.go

```diff
@@ -808,13 +905,237 @@ func (ss *StakerSharesModel) GetStakerSharesFromDB(staker string, strategy strin
+	shareDeltas, ok := ss.shareDeltaAccumulator[blockNumber]
+	if !ok {
+		msg := "delta accumulator was not initialized"
+		ss.logger.Sugar().Errorw(msg, zap.Uint64("blockNumber", blockNumber))
+		return nil, errors.New(msg)
+	}
+	slashes, ok := ss.SlashingAccumulator[blockNumber]
+	if !ok {
+		msg := "slashing accumulator was not initialized"
+		ss.logger.Sugar().Errorw(msg, zap.Uint64("blockNumber", blockNumber))
+		return nil, errors.New(msg)
+	}
+
+	// Ensure that all of our slashes are sorted by transactionIndex asc and logIndex asc
+	slashesByTransaction := make(map[uint64][]*SlashDiff)
+	transactionIndexes := make([]uint64, 0)
+	for _, slash := range slashes {
+		if _, ok := slashesByTransaction[slash.TransactionIndex]; !ok {
+			slashesByTransaction[slash.TransactionIndex] = make([]*SlashDiff, 0)
+			transactionIndexes = append(transactionIndexes, slash.TransactionIndex)
+		}
+		slashesByTransaction[slash.TransactionIndex] = append(slashesByTransaction[slash.TransactionIndex], slash)
+	}
+	slices.Sort(transactionIndexes)
+	orderedSlashes := make([]*SlashDiff, 0)
+	for _, transactionIndex := range transactionIndexes {
+		slices.SortFunc(slashesByTransaction[transactionIndex], func(a, b *SlashDiff) int {
+			return int(a.LogIndex - b.LogIndex)
+		})
+		for _, slash := range slashesByTransaction[transactionIndex] {
+			orderedSlashes = append(orderedSlashes, slash)
+		}
+	}
+
+	// keep track of the running sum of deltas found in this current block, if any
+	netDeltas := make(map[string]*big.Int)
+	slashedDeltas := make([]*StakerShareDeltas, 0)
+	records := make([]*StakerShareDeltas, 0)
+
+	for _, slash := range orderedSlashes {
+		// Iterate over all of the in-memory deltas and find the ones that are before the slashing event
+		//
+		// For the records that DO come before the slashing event, capture their running share sum up to the slashing
+		// event so that we can add it to the existing sum from the database.
+		for _, delta := range shareDeltas {
+			// add the delta to the new list that will contain deposit/withdrawl deltas AND slashes
+			records = append(records, delta)
+
+			if delta.TransactionIndex > slash.TransactionIndex {
+				continue
+			}
+			if delta.TransactionIndex == slash.TransactionIndex && delta.LogIndex > slash.LogIndex {
+				continue
+			}
+			key := getNetDeltaKey(delta.Staker, delta.Strategy)
+			if _, ok := netDeltas[key]; !ok {
+				netDeltas[key] = big.NewInt(0)
+			}
+			shares, success := new(big.Int).SetString(delta.Shares, 10)
+			if !success {
+				return nil, fmt.Errorf("failed to convert shares to big.Int: %s", delta.Shares)
+			}
+			netDeltas[key] = netDeltas[key].Add(netDeltas[key], shares)
+		}
+
+		// Get staker shares up to the current block for all delegated stakers, INCLUDING stakers delegated in this block
+		var stakerShares []*StakerShares
+		var err error
+		if slash.BeaconChain {
+			stakerShares, err = ss.GetStakerSharesFromDB(slash.SlashedEntity, slash.Strategy)
```

Comment:
this also needs to account for if there is a precommitted record for a beacon chain staker